Add `Promoted` to `std.traits`.

tsbockman opened the pull request:
Add a `Promoted!T` trait which evaluates to the type used for (non-unary) arithmetic operations on type `T`. Example: `static assert(is(typeof(Promoted!ubyte == int)));`.

JackStouffer commented on std/traits.d:
```diff
@@ -6490,6 +6509,38 @@ unittest
+    if(isScalarType!T)
+{
+    static if (isFloatingPoint!T)
+        enum int precision = T.mant_dig;
```
`ushort`?

tsbockman replied:
```D
static assert(is(typeof(real.mant_dig) == int));
```

JackStouffer commented on std/traits.d:
```diff
@@ -6524,6 +6575,40 @@ unittest
 }
 
+/**
+Get the type that a scalar type `T` will promote to in arithmetic expressions.
```
IIRC there was a whole five minutes in a DConf talk complaining about these rules. Is there anywhere where these are defined, and if so, could you link to it here?

tsbockman replied:
Link (I will add it to the PR, as well): https://dlang.org/spec/type.html#integer-promotions

For the purposes of my `Promoted` trait, the rules are both simple and sensible:

* Any basic type smaller than `int` is promoted to `int`.

* `dchar` promotes to `uint`.

The problem comes when you mix two different types, like `int` and `uint` - at that point, what happens is the *opposite* of what should happen: the signed type will be promoted to the **un**signed type. This is awful, but required by D's "*C code that compiles in D should do the same thing it did in C*" rule.

The need to warn against implicit signed to unsigned conversions stuff it is what makes `SafeInt` such a pain to use, motivating me to create `SmartInt` which uses better rules.

aG0aep6G commented on std/traits.d:
```diff
@@ -5154,6 +5157,22 @@ unittest
+Detect whether `T` is a scalar type that promotes to an integral type (that is, a built-in
+integral, character, or boolean type).
+ */
+enum bool isFixedPoint(T) = isIntegral!T || isSomeChar!T || isBoolean!T;
```
Is "fixed point" the right term for this? I would expect a number with a fractional part (with *fixed* precision) behind that name.

tsbockman replied:
> Is "fixed point" the right term for this?

I intended it as the opposite of `isFloatingPoint`, since `isFixedPoint == (isScalarType && !isFloatingPoint)`. Do you have a suggestion for a better name?

andralex replied:
Interesting rationale, but the converse of floating point is not fixed point; the latter is a mere alternative for implementing approximations of real numbers. Integrals and their ilk are a distinct matter.

How frequent is this likely to be? Should it include integral `enum` types?

tsbockman replied:
> Interesting rationale, but the converse of floating point is not fixed point...

I am not particularly attached to the `isFixedPoint` name - I'm just waiting for someone to suggest something better: descriptive and meaningfully shorter than the implementation.

> How frequent is this likely to be **[used]**?

When converting functions into templates, people tend to replace the parameter `int a` with `A a` ... `if (isIntegral!A)`. The two are not fully compatible, though, because `int a` will accept `bool`, `char`, and `wchar` as well.

The range of types accepted by `A a` ... `if (isFixedPoint!A)` is much closer to those accepted by `int a`. In my opinion most template functions with `isIntegral` in the constraints are candidates for conversion to `isFixedPoint`, instead.

Of course, the whole thing is just a hack to work around a weakness in the language: the inability to declare a template function in such a way that type deduction works properly when implicit conversions are involved.

> Should it include integral `enum` types?

I considered it unwise to deviate from the behaviour of the existing `isWhatever` traits, so it does the same thing that `isIntegral` does: it gives the same answer for an `enum` type as it would for the `enum`'s base type.

JackStouffer commented:
LGTM

You should email Andrei as these are name additions

andralex commented on std/traits.d:
```diff
@@ -6490,6 +6509,38 @@ unittest
+/**
+Returns the number of full binary digits of precision representable by the scalar type `T`.
+ */
+template precision(T)
```
What would be an example of unifying FP and integral precision?

tsbockman replied:
A floating-point type `F` can losslessly represent all possible values of an integral type `N` if-and-only-if `precision!T >= precision!N`.

This trait is used by `smartOp` to help compute the optimal common numeric type for multi-argument functions like `binary()`.

andralex commented on std/traits.d:
```diff
@@ -6490,6 +6509,38 @@ unittest
+    if (isScalarType!T)
+{
+    static if (isFloatingPoint!T)
+        enum int precision = T.mant_dig;
```
`uint` throughout

tsbockman replied:
```D
static assert(is(typeof(real.mant_dig) == int));
```

Also, D's integer promotion rules generally make `int` the safest and most user-friendly type when the value is guaranteed to be both small and positive.

andralex commented on std/traits.d:
```diff
@@ -6524,6 +6575,41 @@ unittest
 }
 
+/**
+Get the type that a scalar type `T` will $(LINK2 https://dlang.org/spec/type.html#integer-promotions, promote)
```
use the WEB macro instead

wilzbach replied:
no please don't - see [this](https://github.com/dlang/phobos/pull/4411) for the motivation to kill WEB.

andralex replied:
That's been settled in the favor of HTTP or HTTPS.

aG0aep6G replied:
> That's been settled in the favor of HTTP or HTTPS.

No, it hasn't. Everyone agrees that WEB can go away, but I don't agree that HTTP/HTTPS are to be preferred over LINK2. See [my reply](https://github.com/dlang/phobos/pull/4411#issuecomment-226392170) in the linked PR.

andralex commented on std/traits.d:
```diff
@@ -6524,6 +6575,41 @@ unittest
+Get the type that a scalar type `T` will $(LINK2 https://dlang.org/spec/type.html#integer-promotions, promote)
+to in arithmetic expressions.
+*/
+template Promoted(T)
```
nice, approved

JackStouffer commented:
Anyone else?

tsbockman commented:
> Anyone else?

Unresolved:

* A less confusing name is needed for `isFixedPoint`, but none has been suggested yet.
* @andralex has not yet made a decision about `precision`.

andralex commented:
To move forward with this PR, I suggest it be reduced to only add `Promoted`. I don't think `precision` and `isFloatingPoint` (under any name) add enough value. At any rate those can be added later with the appropriate evidence.

andralex commented:
(removing my label after posting my response)

JackStouffer commented:
Ping @tsbockman

tsbockman commented:
@JackStouffer I have removed `precision` and `isFixedPoint` from this PR, added a changelog entry, and tweaked the examples slightly to make it clearer what `Promoted` is for.

JackStouffer commented:
Andrei has approved the addition and the code, no one else has objected, so I'm pulling

JackStouffer commented:
Auto-merge toggled on
